karpenter-operator/controllers/karpenter/karpenter_controller.go

```diff
@@ -128,13 +128,35 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 		return ctrl.Result{}, err
 	}
 	if hcp.DeletionTimestamp != nil {
-		// TODO(alberto): implement deletion. E.g. loop over nodeClaims delete them, wait and delete karpeneter deployment.
 		if controllerutil.ContainsFinalizer(hcp, karpenterFinalizer) {
+			nodePoolList := &unstructured.UnstructuredList{}
+			nodePoolList.SetAPIVersion("karpenter.sh/v1")
+			nodePoolList.SetKind("NodePool")
+			// TODO(maxcao13): Import and register the openshift mirrored kubernetes-sigs/karpenter types
+			if err := r.GuestClient.List(ctx, nodePoolList, client.InNamespace(r.Namespace)); err != nil {
+				return ctrl.Result{}, fmt.Errorf("failed to list NodePools: %w", err)
+			}
+
+			for _, nodePool := range nodePoolList.Items {
+				// If we still get the NodePool, but it's already marked as terminating, we don't need to call Delete again
+				if !nodePool.GetDeletionTimestamp().IsZero() {
+					continue
+				}
+				if err := r.GuestClient.Delete(ctx, &nodePool); err != nil {
+					return ctrl.Result{}, fmt.Errorf("failed to delete NodePool: %w", err)
+				}
+			}
+			// Wait until all NodePools are deleted before removing the finalizer
+			if len(nodePoolList.Items) > 0 {
```

Comment:
This will cause the nodePools to immediately be removed from etcd and signal background deletion for the nodeClaims, so we also need to wait for all nodeClaims to go away, agree?

Reply:
Ah right, makes sense, I was assuming the default behaviour was foreground deletion. Would it make sense to just set foreground deletion specifically here?

Reply:
Lets codify nodeClaim wait for deletion explicitly so we ensure our criteria for infra removal can't be broken by a change/bug in the karpenter ownership chain. If we want to signal the request as foreground additional that seems fine as well.